File: app/src/main/java/com/doyou/cvc/release/listener/SettingsSystemObserver.java
package com.doyou.cvc.release.listener;

import android.content.ContentResolver;
import android.database.ContentObserver;
import android.net.Uri;
import android.os.Handler;
import android.provider.Settings;

import java.lang.ref.WeakReference;

/**
 * @autor hongbing
 * @date 2019-10-08
 */
public class SettingsSystemObserver extends ContentObserver {

    public String mName;
    private WeakReference<Callback> mCallback;

    public interface Callback {
        void mtcSettingsSystemChanged(SettingsSystemObserver observer);
    }

    public SettingsSystemObserver(Handler handler) {
        super(handler);
    }

    public void setCallback(Callback callback) {
        mCallback = (callback == null) ? null : new WeakReference<Callback>(callback);
    }

    public Callback getCallback() {
        return (mCallback == null) ? null : mCallback.get();
    }

    public void register(ContentResolver contentResolver, String name) {
        Uri uri = Settings.System.getUriFor(name);
        contentResolver.registerContentObserver(uri, false, this);
        mName = name;
    }

    public void unregister(ContentResolver contentResolver) {
        contentResolver.unregisterContentObserver(this);
    }

    public void onChange(boolean selfChange) {
        Callback callback = getCallback();
        if (callback != null) {
            callback.mtcSettingsSystemChanged(this);
        }
    }
}

File: app/src/main/java/com/doyou/cvc/activity/listener/SettingsSystemObserver.kt
package com.doyou.cvc.activity.listener

import android.content.ContentResolver
import android.database.ContentObserver
import android.os.Handler
import android.provider.Settings
import java.lang.ref.WeakReference

/**
 * @autor hongbing
 * @date 2019-10-08
 */
class SettingsSystemObserver(handler: Handler?) : ContentObserver(handler) {
    var mName: String? = null
    private var mCallback: WeakReference<Callback>? = null

    interface Callback {
        fun mtcSettingsSystemChanged(observer: SettingsSystemObserver?)
    }

    var callback: Callback?
        get() = if (mCallback == null) null else mCallback!!.get()
        set(callback) {
            mCallback = if (callback == null) null else WeakReference(callback)
        }

    fun register(contentResolver: ContentResolver, name: String?) {
        val uri = Settings.System.getUriFor(name)
        contentResolver.registerContentObserver(uri, false, this)
        mName = name
    }

    fun unregister(contentResolver: ContentResolver) {
        contentResolver.unregisterContentObserver(this)
    }

    override fun onChange(selfChange: Boolean) {
        val callback = callback
        callback?.mtcSettingsSystemChanged(this)
    }
}